File: src/com/jepl/lang/Interpreter.java
package com.jepl.lang;

import com.ezylang.evalex.EvaluationException;
import com.ezylang.evalex.Expression;
import com.ezylang.evalex.parser.ParseException;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.jepl.lang.libraries.Function;

import java.awt.*;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.List;

public class Interpreter {
    public static int block = 0;
    public static void interpret(String input){

        JsonArray code = JsonParser.parseString(input).getAsJsonArray();
        Scanner scanner = new Scanner(System.in);
        for(int i = 0; i < code.size(); i++){
            block = i;
            JsonObject o = code.get(i).getAsJsonObject();
            String name = o.get("name").getAsString();
            JsonArray args = o.get("args").getAsJsonArray();
            if(name.equals("print")){
                List<String> arg = getArgs(args);
                for(String a: arg){
                    System.out.print(a+" ");
                }
            }else if (name.equals("exit")){
                System.exit(args.get(0).getAsInt());
            }else if(name.equals("println")){
                List<String> arg = getArgs(args);
                for(String a: arg){
                    System.out.println(a);
                }
            }else if(name.equals("function")){
                String body = o.get("body").getAsJsonArray().toString();
                String funName = getArgs(args).get(0);
                Values.addFun(funName, body);
            }else if(name.equals("jump")){
                final List<String> arg = getArgs(args);
                String funName = arg.get(0);
                int argc = 0;
                if(arg.size() > 1) {
                    for (int a = 1; a < arg.size(); a++) {
                        argc = a;
                        Values.addVar("arg" + a, arg.get(a));
                    }
                }
                Values.vars.put("argc", argc);
                try {
                    Interpreter.interpret(Values.getFun(funName));
                }catch (IllegalStateException e){
                    throw new JEPLException(new NoSuchFieldException(funName + " function is not exist"));
                }
                Values.vars.remove("argc");
                if(arg.size() > 1) {
                    for (int a = 1; a < arg.size(); a++) {
                        Values.vars.remove("arg" + a);
                    }
                }
            }else if(name.equals("for")){
                String body = o.get("body").getAsJsonArray().toString();
                List<String> arg = getArgs(args);
                int from, to;
                try {
                    from = Integer.parseInt(arg.get(0));
                    to = Integer.parseInt(arg.get(1));
                }catch (NumberFormatException e){
                    throw new JEPLException(new RuntimeException(e));
                }
                for(int x = from; x <= to; x++){
                    Values.addVar("index", String.valueOf(x));
                    Interpreter.interpret(body);
                }
                Values.vars.remove("index");
            } else if (name.equals("variable")) {
                List<String> arg = getArgs(args);
                String n = arg.get(0);
                String v = arg.get(1);
                Values.addVar(n, v);
            }else if(name.equals("input")) {
                List<String> arg = getArgs(args);
                String text = arg.get(1);
                String var = arg.get(0);
                System.out.println(text);
                String inp = scanner.nextLine();
                Values.addVar(var, inp);
            }else if(name.equals("rmvar")) {
                List<String> arg = getArgs(args);
                String var = arg.get(0);
                if (Values.vars.containsKey(var)) {
                    Values.vars.remove(var);
                } else {
                    throw new JEPLException(new RuntimeException("Warning [block "+i+" - variable "+var+" not exists!"));
                }
            }else if(name.equals("include")) {
                List<String> arg = getArgs(args);
                String filename = arg.get(0);
                try {
                    Interpreter.interpret(Files.readString(Path.of(filename)));
                } catch (IOException e) {
                    throw new JEPLException(new FileNotFoundException("Error [block "+block + 1+" - file "+filename+" is not exists!"));
                }
            }else if(name.equals("while")) {
                List<String> arg = getArgs(args);
                final String body = o.get("body").getAsJsonArray().toString();
                boolean condition = Boolean.parseBoolean(arg.get(0));
                while (condition) {
                    Interpreter.interpret(body);
                    arg = getArgs(args);
                    condition = Boolean.parseBoolean(arg.get(0));
                    if (!condition) {
                        break;
                    }
                }
            }else if(name.equals("if")) {
                List<String> arg = getArgs(args);
                boolean condition = Boolean.parseBoolean(arg.get(0));
                String body = o.get("body").getAsJsonArray().toString();
                if (condition) {
                    Interpreter.interpret(body);
                }
            }else if(name.equals("import")) {
                List<String> arg = getArgs(args);
                for(String s: arg){
                    if(!Values.libs.containsKey(s)) {
                        Values.addLib(s);
                    }else {
                        throw new JEPLException(new RuntimeException("Library <"+s+"> already defined!"));
                    }
                }
            }else if(name.equals("eval")) {
                List<String> arg = getArgs(args);
                String eval = arg.get(0);
                Interpreter.interpret(eval);
            }else if(name.equals("runtime_execute")) {
                List<String> arg = getArgs(args);
                String command = arg.get(0);
                try {
                    Runtime.getRuntime().exec(command);
                } catch (IOException e) {
                    throw new JEPLException(new RuntimeException(e));
                }
            }else if(name.equals("unload")) {
                List<String> arg = getArgs(args);
                for(String s: arg){
                    Values.libs.remove(s);
                }
            }else if(name.equals("getProperty")) {
                final List<String> arg = getArgs(args);
                String prop = arg.get(0);
                String to = arg.get(1);
                Values.addVar(to, System.getProperty(prop));
            }else if(name.equals("importJar")){
                final List<String> arg = getArgs(args);
                String path = System.getProperty("user.dir")+"/"+arg.get(0);
                String addressLib = arg.get(1);
                try {
                    URLClassLoader child = new URLClassLoader(
                            new URL[] { new URL("file:" + path) },
                            Interpreter.class.getClassLoader()
                    );
                    Class<?> classToLoad = Class.forName(addressLib + ".Lib", true, child);
                    Method method = classToLoad.getDeclaredMethod("invoke");
                    Object instance = classToLoad.getDeclaredConstructor().newInstance();
                    HashMap<String, Function> lib = (HashMap<String, Function>) method.invoke(instance);
                    Values.libs.put(addressLib,  lib);

                } catch (MalformedURLException | ClassNotFoundException | NoSuchMethodException | InstantiationException | IllegalAccessException |
                         InvocationTargetException e) {
                    throw new RuntimeException(e);
                }
            }else{
                List<String> arg = getArgs(args);
                int index = 0;
                for(Map.Entry<String, HashMap<String, Function>> lib: Values.libs.entrySet()){
                    if(lib.getValue().containsKey(name)) {
                        lib.getValue().get(name).invoke(arg);
                        break;
                    }
                    index++;
                }
                if (index == Values.libs.size()){
                    throw new JEPLException(new NoSuchFieldException("Cannot find method "+name));
                }
            }
        }
    }

    public static List<String> getArgs(JsonArray a){
        List<String> out = new ArrayList<>();
        for(JsonElement n: a){
            Expression e = getExpression(n);
            try {
                out.add(e.evaluate().getStringValue());
            } catch (EvaluationException | ParseException ex) {
                throw new JEPLException(new RuntimeException(ex.getMessage()));
            }
        }
        return out;
    }

    public static Expression getExpression(JsonElement n) {
        Expression e = new Expression(n.getAsString());
        e.with("jeplversion", Runner.langVersion);
        for(int i = 0; i < Values.vars.size(); i++){
            for(Map.Entry<String, Object> map: Values.vars.entrySet()) {
                Object val;
                try {
                    val = Integer.parseInt(map.getValue().toString());
                }catch (NumberFormatException ev){
                    val = map.getValue();
                }
                e.with(map.getKey(), val);
            }
        }
        return e;
    }
}

File: src/com/jepl/lang/Interpreter.kt
package com.jepl.lang

import com.google.gson.JsonParser
import com.jepl.lang.Expressions.Companion.checkArguments
import com.jepl.lang.Expressions.Companion.getArgs
import java.io.File
import java.io.FileNotFoundException
import java.net.URI
import java.net.URLClassLoader
import kotlin.system.exitProcess
import com.jepl.lang.libraries.Function
import com.jepl.lang.libraries.Library

class Interpreter {
    companion object {
        var block: Int = 0

        fun interpret(input: String) {
            val code = JsonParser.parseString(input).asJsonArray
            for (i in 0..<code.size()) {
                block = i

                val o = code.get(i).asJsonObject
                val name = o.get("name").asString
                val _args = o.get("args").asJsonArray
                when (name) {
                    "print" -> {
                        val args = getArgs(_args)
                        for (arg in args) {
                            print(arg)
                        }
                    }

                    "println" -> {
                        val args = getArgs(_args)
                        for (arg in args) {
                            println(arg)
                        }
                    }

                    "exit" -> {
                        checkArguments(1, getArgs(_args), o)
                        exitProcess(_args.get(0).asInt)
                    }

                    "function" -> {
                        val body = o.get("body").asJsonArray.toString()
                        checkArguments(1, getArgs(_args), o)
                        val funName = getArgs(_args)[0]
                        Values.addFun(funName, body)
                    }

                    "jump" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val funName = args[0]
                        var argc = 0
                        if (args.size > 1) {
                            for (a in 1..<args.size) {
                                argc += 1
                                Values.addVar("arg$a", args[a])
                            }
                        }
                        Values.vars["argc"] = argc
                        try {
                            interpret(Values.getFun(funName).toString())
                        } catch (e: IllegalStateException) {
                            throw JEPLException(NoSuchFieldException(e.message), o)
                        }
                        Values.vars.remove("argc")
                        if (args.size > 1) {
                            for (a in 1..<args.size) {
                                Values.vars.remove("arg$a")
                            }
                        }
                    }

                    "for" -> {
                        val body = o.get("body").asString
                        val args = getArgs(_args)
                        checkArguments(2, args, o)
                        val from: Int;
                        val to: Int
                        try {
                            from = Integer.parseInt(args[0]); to = Integer.parseInt(args[1])
                        } catch (e: NumberFormatException) {
                            throw JEPLException(e, o)
                        }
                        for (iter in from..to) {
                            Values.addVar("index", iter)
                            interpret(body)
                        }
                        Values.vars.remove("index")
                    }

                    "variable" -> {
                        val args = getArgs(_args)
                        checkArguments(2, args, o)
                        val varName = args[0];
                        val varContent = args[1]
                        Values.addVar(varName, varContent)
                    }

                    "input" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val text = args[0];
                        val outTo = args[1]
                        println(text)
                        val inputs: String? = readlnOrNull()
                        if (inputs != null) {
                            Values.addVar(outTo, inputs)
                        }
                    }

                    "rmvar" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val varName = args[0]
                        Values.vars.remove(varName)
                    }

                    "include" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val filename = args[0]
                        try {
                            interpret(File(filename).readText())
                        } catch (e: FileNotFoundException) {
                            throw JEPLException(
                                FileNotFoundException("Error [block${block + 1}] - file $filename not found either not exists!"),
                                o
                            )
                        }
                    }

                    "while" -> {
                        var args = getArgs(_args)
                        checkArguments(1, args, o)
                        val body = o["body"].asJsonArray.toString()
                        var condition = args[0].toBoolean()
                        while (condition) {
                            interpret(body)
                            args = getArgs(_args)
                            condition = args[0].toBoolean()
                            if (!condition) {
                                break
                            }
                        }
                    }

                    "if" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val condition = args[0].toBoolean()
                        val body = o["body"].asJsonArray.toString()
                        if (condition) {
                            interpret(body)
                        }
                    }

                    "import" -> {
                        val args = getArgs(_args)
                        for (arg in args) {
                            when (Values.libs.containsKey(arg)) {
                                true -> {
                                    throw JEPLException(RuntimeException("Library <$arg> already defined!"), o)
                                }

                                false -> {
                                    Values.addLib(arg)
                                }
                            }
                        }
                    }

                    "eval" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val codeIn = args[0]
                        interpret(codeIn)
                    }

                    "runtime_execute" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val cmd = args[0]
                        try {
                            Runtime.getRuntime().exec(cmd)
                        } catch (e: RuntimeException) {
                            throw JEPLException(e, o)
                        }
                    }

                    "unload" -> {
                        val args = getArgs(_args)
                        checkArguments(1, args, o)
                        val lib = args[0]
                        when (Values.libs.containsKey(lib)) {
                            true -> Values.libs.remove(lib)
                            false -> throw JEPLException(RuntimeException("Library $lib wasn't imported!"), o)
                        }
                    }

                    "get_property" -> {
                        val args = getArgs(_args)
                        checkArguments(2, args, o)
                        val property = args[0]
                        val variable = args[1]
                        Values.addVar(variable, System.getProperty(property))
                    }

                    "set_property" -> {
                        val args = getArgs(_args)
                        checkArguments(2, args, o)
                        val property = args[0];
                        val value = args[1]
                        System.setProperty(property, value)
                    }

                    "import_jar" -> {
                        val args = getArgs(_args)
                        checkArguments(2, args, o)
                        val path = "${System.getProperty("user.dir")}\\${args[0].replace("/", "\\")}"
                        val addr = args[1]
                        try {
                            val urls = arrayOf(URI("file:$path").toURL())
                            val loader = URLClassLoader(urls, this::class.java.classLoader)
                            val classOf = Class.forName("$addr.Lib", true, loader)
                            val method = classOf.getDeclaredMethod("invoke")
                            val instance = classOf.getDeclaredConstructor().newInstance()

                            @Suppress("UNCHECKED_CAST")
                            val lib: HashMap<String, Function> = method.invoke(instance) as HashMap<String, Function>
                            Values.libs[addr] = lib
                        } catch (e: Exception) {
                            throw JEPLException(e, o)
                        }
                    }

                    "neg" -> {
                        val args = getArgs(_args)
                        val v = args[0]
                        val value = args[1]
                        if (Values.vars.containsKey(v))
                            Values.vars[v] = Integer.parseInt(Values.vars[v].toString()) + Integer.parseInt(value)
                    }

                    "assert" -> {
                        checkArguments(1, getArgs(_args), o)
                        val condition = getArgs(_args)[0]
                        if (!condition.toBoolean())
                            throw JEPLException(AssertionError("Condition is false"), o)
                    }

                    "preprocessor" -> {
                        val args = getArgs(_args)
                        checkArguments(2, args, o)
                        val `in` = args[0]
                        val of = args[1]
                        val generatedLib = object : Library {
                            override fun invoke(): HashMap<String, Function> {
                                val out = HashMap<String, Function>()
                                out[of] = object : Function {
                                    override fun invoke(args: List<String>) {
                                        val sb = StringBuilder()
                                        sb.append('[')
                                        for (arg in args) {
                                            if (arg != args.last())
                                                sb.append("\"\\\"").append(arg).append("\\\"\"").append(", ")
                                            else
                                                sb.append("\"\\\"").append(arg).append("\\\"\"")
                                        }
                                        sb.append("]")
                                        interpret("[{\"name\": \"${`in`}\", args: ${sb}}]")
                                    }
                                }
                                return out
                            }
                        }
                        Values.libs["$!generated$?lib!${`in`}"] = generatedLib.invoke()
                    }

                    else -> {
                        val args = getArgs(_args)
                        var index = 0

                        for (lib in Values.libs.iterator()) {
                            if (lib.value.containsKey(name)) {
                                lib.value[name]?.invoke(args)
                                break
                            }
                            ++index
                        }

                        when (index == Values.libs.size) {
                            true -> throw JEPLException(NoSuchFieldException("field $name wasn't found"), o)
                            false -> continue
                        }
                    }
                }
            }
        }

    }
}